### ISSUE TITLE
OS X Compilation

### ISSUE THREAD

**Issue by traversaro** [ocra-recipes/ocra-recipes#4]
I tried to compile on OS X, and the issue is that the link dependencies between the different libraries (`ocra_optim`, `ocra_control`, `quadprog`, `gocra`) are not declared in the `CMakeLists.txt`. This is ok on Linux but not on OS X, because of the difference in the linking between the two different OS.

**Comment by traversaro** [comment]
I try to add in a naive way the interlibrary dependencies in https://github.com/traversaro/ocra-recipes/commit/fdbaee36c3f69565a78e80d846974ec7b028d115 , but apparently this breaks the logic of the exported `***Config.cmake` files.  
When I try to compile `ocra-wbi-plugins` I got this error: 

```
[traversaro@mbp ~/src/robotology/codyco-superbuild/build/main/ocra-wbi-plugins (master)]$ cmake .
-- Using iCub from build tree
-- Found YCM: /Users/traversaro/src/robotology/codyco-superbuild/build/install/share/YCM (found version "0.2.0~20151028.3+gitaae056b")
-- Boost version: 1.59.0
-- Found the following Boost libraries:
--   system
--   filesystem
-- Eigen found (include: /usr/local/Cellar/eigen/3.2.6/include/eigen3)
-- Using YARP from build tree: no file will we be installed into path.d folder, you need to set YARP_DATA_DIRS environment variable to /Users/traversaro/src/robotology/yarp/build/share/yarp:/Users/traversaro/src/robotology/icub-main/build/share/icub:/Users/traversaro/src/robotology/codyco-superbuild/build/install/share/codyco:/Users/traversaro/src/robotology/codyco-superbuild/build/main/ocra-wbi-plugins/share/codyco
CMake Error at libs/ocraWbiPlugins/CMakeLists.txt:4 (find_package):
  Found package configuration file:

    /Users/traversaro/src/robotology/codyco-superbuild/build/install/share/cmake/wocra/WOCRAConfig.cmake

  but it set WOCRA_FOUND to FALSE so package "WOCRA" is considered to be NOT
  FOUND.  Reason given by package:

  The following imported targets are referenced, but are missing:
  ocra_control quadprog
```

This is due to the fact that every part of ocra-recepies is exported as a different CMake package, and so the `WOCRAConfig.cmake` is dependent on the `ocra_control` and `quadprog` library, but this imported targets are defined in a different CMake project that should be properly found by `WOCRAConfig.cmake` itself.  

For dealing with this type of situation we developed a specific CMake macro at IIT ( https://github.com/robotology/ycm/blob/master/modules/InstallBasicPackageFiles.cmake ). If you are happy with that it would be easy to migrate the `ocra-recepies` to use `InstallBasicPackageFiles`. 

Using `InstallBasicPackageFiles` the main drawbacks with respect to the current setup would be: 
- To fully exploit the power of define the header interfaces of the libraries, we should depend on CMake 2.8.11 (i.e. dropping support for the default CMake version in Ubuntu 12.04). 
- We will not be able to do `export(PACKAGE ...)`. This is anyway quite a tricky command, that could cause conflicts if different version of the same library are installed in the same machine. 

cc @ahoarau @rlober any opinions?

**Comment by rlober** [comment]
Merry Christmas eve @traversaro 
I am totally fine with changing how we install ocra. It seems like a nice option to keep the packages separate for the sake of modularity (this is more of a long term concern really). We should definitely add the uninstall target too.

@ahoarau and I can take a look at these issues next week and try to make some progress. I know he is working on a [Travis build for ocra-wbi-plugins](https://github.com/ocra-recipes/ocra-wbi-plugins/commit/52bf22d72075655f297b95131e6c04aea48691b5) but I believe he is having some issues with the [YCM and superbuild dependencies](https://travis-ci.org/ocra-recipes/ocra-wbi-plugins/jobs/98377272).

**Comment by traversaro** [comment]
> It seems like a nice option to keep the packages separate for the sake of modularity (this is more of a long term concern really). 

Keeping the package separated is definitely possible, but if a package (let's say `gocra`) uses `ocra` data structures in its public interface (i.e. if `ocra` headers are included in the public headers of `gocra`) the appropriate `find_package` should be added in the `GOCRAConfig.cmake`. See for example how we do in `iDynTree` : https://github.com/robotology/idyntree/blob/master/iDynTreeConfig.cmake.in#L15 .

**Comment by traversaro** [comment]
And Merry Christmas! :christmas_tree:

**Comment by rlober** [comment]
Closed by @ahoarau in 84af60642bf300f767738bbc7a6ce8896beaa1d1 & 8ce3226d250c54e5bfb6c6717dd9dde2a05367f8

:+1: @ahoarau and @traversaro
